File: backend/app/api/routes/insights.py
"""
Read-only insight routes for queue, review state, and audit artifacts.
"""
from __future__ import annotations

import json
from typing import Any, Literal, Optional

from fastapi import APIRouter, HTTPException, Query
from fastapi.responses import Response
from pydantic import BaseModel, Field

from ...services import read_only as read_only_service
from ...services.track_service import get_track_by_id
from modules import enrichment_apply

router = APIRouter(tags=["insights"])


ReviewStatus = Literal["approved", "rejected", "deferred"]


class ReviewStateItem(BaseModel):
    track_id: int
    review_status: ReviewStatus
    updated_at: Optional[str] = None


class ReviewStateResponse(BaseModel):
    items: dict[str, ReviewStateItem]
    approved: list[int]
    rejected: list[int]
    deferred: list[int]
    counts: dict[str, int]
    approved_high_count: int = 0
    approved_medium_count: int = 0
    rejected_by_reason: dict[str, int] = Field(default_factory=dict)
    queue_total: int = 0
    updated_at: Optional[str] = None


class ReviewSummaryResponse(BaseModel):
    pending_count: int
    approved_count: int
    rejected_count: int
    deferred_count: int
    approved_high_count: int
    approved_medium_count: int
    rejected_by_reason: dict[str, int] = Field(default_factory=dict)
    last_updated: Optional[str] = None


class EnrichmentQueueResponse(BaseModel):
    items: list[dict[str, Any]]
    counts: dict[str, dict[str, int]]
    limit: int
    offset: int
    total: int


class ReviewActionResponse(BaseModel):
    track_id: int
    review_status: ReviewStatus
    state: ReviewStateResponse


class EnrichmentApplyResponse(BaseModel):
    root: str
    db_path: str
    state_path: str
    log_path: str
    dry_run: bool
    approved_seen: int
    proposed_count: int
    applied_count: int
    skipped_count: int
    changes: list[dict[str, Any]]
    skipped: list[dict[str, Any]]


@router.get("/enrichment/queue", response_model=EnrichmentQueueResponse)
async def get_enrichment_queue(
    action: Optional[str] = Query(
        default=None,
        description="Filter by action suggestion (auto_candidate, review, ignore).",
    ),
    confidence: Optional[str] = Query(
        default=None,
        description="Filter by confidence tier (HIGH, MEDIUM, LOW).",
    ),
    limit: int = Query(default=100, ge=1, le=1000),
    offset: int = Query(default=0, ge=0),
) -> EnrichmentQueueResponse:
    payload = read_only_service.load_enrichment_queue(
        action=action,
        confidence=confidence,
        limit=limit,
        offset=offset,
    )
    return EnrichmentQueueResponse(**payload)


@router.get("/enrichment/review/state", response_model=ReviewStateResponse)
async def get_enrichment_review_state() -> ReviewStateResponse:
    return ReviewStateResponse(**read_only_service.load_review_state())


@router.get("/enrichment/review/export")
async def export_enrichment_review_state() -> Response:
    state = read_only_service.load_review_state()
    payload = {
        "approved": state.get("approved", []),
        "rejected": state.get("rejected", []),
        "deferred": state.get("deferred", []),
        "counts": state.get("counts", {}),
        "updated_at": state.get("updated_at"),
    }
    body = json.dumps(payload, indent=2, sort_keys=True)
    headers = {
        "Content-Disposition": 'attachment; filename="enrichment_review_state.json"',
    }
    return Response(content=body, media_type="application/json", headers=headers)


@router.get("/enrichment/review/summary", response_model=ReviewSummaryResponse)
async def get_enrichment_review_summary() -> ReviewSummaryResponse:
    return ReviewSummaryResponse(**read_only_service.build_review_summary())


def _review_action(track_id: int, review_status: ReviewStatus) -> ReviewActionResponse:
    track = get_track_by_id(track_id)
    if track is None:
        raise HTTPException(status_code=404, detail=f"Track {track_id} not found")
    queue_item = read_only_service.lookup_enrichment_queue_item(track.filepath)
    if queue_item is None:
        raise HTTPException(
            status_code=404,
            detail=f"Track {track_id} is not present in the enrichment queue",
        )
    try:
        state = read_only_service.set_review_state(track_id, review_status, queue_item=queue_item)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    return ReviewActionResponse(
        track_id=track_id,
        review_status=review_status,
        state=ReviewStateResponse(**state),
    )


@router.post("/enrichment/review/{track_id}/approve", response_model=ReviewActionResponse)
async def approve_enrichment_review(track_id: int) -> ReviewActionResponse:
    return _review_action(track_id, "approved")


@router.post("/enrichment/review/{track_id}/reject", response_model=ReviewActionResponse)
async def reject_enrichment_review(track_id: int) -> ReviewActionResponse:
    return _review_action(track_id, "rejected")


@router.post("/enrichment/review/{track_id}/defer", response_model=ReviewActionResponse)
async def defer_enrichment_review(track_id: int) -> ReviewActionResponse:
    return _review_action(track_id, "deferred")


@router.post("/enrichment/apply-approved/dry-run", response_model=EnrichmentApplyResponse)
async def dry_run_enrichment_apply_approved() -> EnrichmentApplyResponse:
    result = enrichment_apply.build_approved_enrichment_plan(read_only_service.get_library_root())
    return EnrichmentApplyResponse(**result)


@router.post("/enrichment/apply-approved/apply", response_model=EnrichmentApplyResponse)
async def apply_enrichment_apply_approved(confirm: bool = Query(default=False)) -> EnrichmentApplyResponse:
    if not confirm:
        raise HTTPException(status_code=400, detail="confirm=true is required to apply approved enrichment updates")
    result = enrichment_apply.apply_approved_enrichment(read_only_service.get_library_root(), apply=True)
    return EnrichmentApplyResponse(**result)


@router.get("/audit/latest")
async def get_latest_audit() -> dict[str, Any]:
    report = read_only_service.load_latest_audit_json()
    if report is None:
        return {"available": False}
    return report

File: frontend/src/api/insights.ts
import { apiFetch } from './client'

export type ReviewStatus = 'approved' | 'rejected' | 'deferred' | 'pending'

export interface EnrichmentQueueItem {
  filepath?: string | null
  track_id?: number | null
  provider?: string | null
  confidence?: 'HIGH' | 'MEDIUM' | 'LOW' | 'UNKNOWN' | null
  action_suggestion?: 'auto_candidate' | 'review' | 'ignore' | string | null
  score?: number | null
  review_status?: ReviewStatus | null
  review_updated_at?: string | null
  query?: Record<string, unknown> | null
  best_match?: Record<string, unknown> | null
  [key: string]: unknown
}

export interface EnrichmentQueueResponse {
  items: EnrichmentQueueItem[]
  counts: Record<string, Record<string, number>>
  limit: number
  offset: number
  total: number
}

export interface ReviewStateItem {
  track_id: number
  review_status: Exclude<ReviewStatus, 'pending'>
  updated_at: string | null
}

export interface ReviewStateResponse {
  items: Record<string, ReviewStateItem>
  approved: number[]
  rejected: number[]
  deferred: number[]
  counts: Record<string, number>
  approved_high_count: number
  approved_medium_count: number
  rejected_by_reason: Record<string, number>
  queue_total: number
  updated_at: string | null
}

export interface ReviewSummaryResponse {
  pending_count: number
  approved_count: number
  rejected_count: number
  deferred_count: number
  approved_high_count: number
  approved_medium_count: number
  rejected_by_reason: Record<string, number>
  last_updated: string | null
}

export interface ApplyApprovedChange {
  track_id: number
  filepath: string
  fields: string[]
  before: Record<string, unknown>
  after: Record<string, unknown>
  confidence: string
  provider: string
  score: number | null
  review_status: string
}

export interface ApplyApprovedResponse {
  root: string
  db_path: string
  state_path: string
  log_path: string
  dry_run: boolean
  approved_seen: number
  proposed_count: number
  applied_count: number
  skipped_count: number
  changes: ApplyApprovedChange[]
  skipped: Array<Record<string, unknown>>
}

export function fetchLatestAudit(): Promise<Record<string, unknown>> {
  return apiFetch.get<Record<string, unknown>>('/audit/latest')
}

export function fetchEnrichmentQueue(params?: {
  action?: 'auto_candidate' | 'review' | 'ignore' | null
  confidence?: 'HIGH' | 'MEDIUM' | 'LOW' | null
  limit?: number
  offset?: number
}): Promise<EnrichmentQueueResponse> {
  const query = new URLSearchParams()
  if (params?.action) query.set('action', params.action)
  if (params?.confidence) query.set('confidence', params.confidence)
  if (params?.limit != null) query.set('limit', String(params.limit))
  if (params?.offset != null) query.set('offset', String(params.offset))
  const suffix = query.toString() ? `?${query.toString()}` : ''
  return apiFetch.get<EnrichmentQueueResponse>(`/enrichment/queue${suffix}`)
}

export function fetchReviewState(): Promise<ReviewStateResponse> {
  return apiFetch.get<ReviewStateResponse>('/enrichment/review/state')
}

export function fetchReviewSummary(): Promise<ReviewSummaryResponse> {
  return apiFetch.get<ReviewSummaryResponse>('/enrichment/review/summary')
}

export function approveReview(trackId: number) {
  return apiFetch.post<{ track_id: number; review_status: 'approved'; state: ReviewStateResponse }>(
    `/enrichment/review/${trackId}/approve`,
    {},
  )
}

export function rejectReview(trackId: number) {
  return apiFetch.post<{ track_id: number; review_status: 'rejected'; state: ReviewStateResponse }>(
    `/enrichment/review/${trackId}/reject`,
    {},
  )
}

export function deferReview(trackId: number) {
  return apiFetch.post<{ track_id: number; review_status: 'deferred'; state: ReviewStateResponse }>(
    `/enrichment/review/${trackId}/defer`,
    {},
  )
}

export function dryRunApplyApproved(): Promise<ApplyApprovedResponse> {
  return apiFetch.post<ApplyApprovedResponse>('/enrichment/apply-approved/dry-run', {})
}

export function applyApproved(confirm = false): Promise<ApplyApprovedResponse> {
  const suffix = confirm ? '?confirm=true' : ''
  return apiFetch.post<ApplyApprovedResponse>(`/enrichment/apply-approved/apply${suffix}`, {})
}
